<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>nullranges: Generation of nullranges via bootstrapping or covariate matching • Bioc2022nullranges</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="nullranges: Generation of nullranges via bootstrapping or covariate matching">
<meta property="og:description" content="Bioc2022nullranges">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Bioc2022nullranges</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/nullranges.html">nullranges: Generation of nullranges via bootstrapping or covariate matching</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/EricSDavis/Bioc2022nullranges" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>nullranges: Generation of nullranges via
bootstrapping or covariate matching</h1>
                        <h4 data-toc-skip class="author">Wancen Mu, Eric
S. Davis</h4>
            
            <h4 data-toc-skip class="date">July 2022</h4>
      
      
      <div class="hidden name"><code>nullranges.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h3>
<p>The nullranges package contains functions for generation of feature
sets (genomic regions) for exploring the null hypothesis of overlap or
colocalization of two observed feature sets. The package has two
approaches for generating null feature sets, <em>matching</em> and
<em>bootstrapping</em>:</p>
<ul>
<li>
<p>Matched subsampling with matchRanges:</p>
<p>Subsampling from a pool of features, but controlling for certain
characteristics.</p>
</li>
<li><p>Bootstrapping with bootRanges:</p></li>
</ul>
<p>In this workshop, we will demonstrate the two major approaches for
generating null hypothesis genomic ranges with the
<code>nullranges</code> package.</p>
</div>
<div class="section level3">
<h3 id="background-and-other-methods">Background and other methods<a class="anchor" aria-label="anchor" href="#background-and-other-methods"></a>
</h3>
<p>Suppose we want to examine the significance of overlaps of genomic
sets of features <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span>. To test the significance of this
overlap, we calculate the overlap expected under the null by generating
a null feature set <span class="math inline">\(y'\)</span>
(potentially many times). The null features in <span class="math inline">\(y'\)</span> may be characterized by:</p>
<ol style="list-style-type: decimal">
<li>Drawing from a larger pool <span class="math inline">\(z\)</span>
(<span class="math inline">\(y' \subset z\)</span>), such that <span class="math inline">\(y\)</span> and <span class="math inline">\(y'\)</span> have a similar distribution over
one or more covariates. This is the “matching” case. Note that the
features in <span class="math inline">\(y'\)</span> are original
features, just drawn from a different pool than y.</li>
<li>Generating a new set of genomic features <span class="math inline">\(y'\)</span>, constructing them from the
original set <span class="math inline">\(y\)</span> by selecting blocks
of the genome with replacement, i.e. such that features can be sampled
more than once. This is the “bootstrapping” case. Note that, in this
case, <span class="math inline">\(y'\)</span> is an artificial
feature set, although the re-sampled features can retain covariates such
as score from the original feature set <span class="math inline">\(y\)</span>.</li>
</ol>
<div class="section level4">
<h4 id="in-other-words">In other words<a class="anchor" aria-label="anchor" href="#in-other-words"></a>
</h4>
<ol style="list-style-type: decimal">
<li>Matching – drawing from a pool of features but controlling for
certain characteristics</li>
<li>Bootstrapping – placing a number of artificial features in the
genome but controlling for their spatial distribution</li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="releated-work">Releated work<a class="anchor" aria-label="anchor" href="#releated-work"></a>
</h3>
<p>For general considerations of generation of null feature sets or
segmentation for enrichment or colocalization analysis, consider the
papers of <span class="citation">De, Pedersen, and Kechris
(2014)</span>, <span class="citation">Haiminen, Mannila, and Terzi
(2007)</span>, <span class="citation">Huen and Russell (2010)</span>,
and <span class="citation">Kanduri et al. (2019)</span> (with links in
references below). Other Bioconductor packages that offer randomization
techniques for enrichment analysis include: <a href="https://cran.r-project.org/web/packages/MatchIt/vignettes/MatchIt.html" class="external-link">MatchIt</a>
<a href="https://bioconductor.org/packages/LOLA" class="external-link">LOLA</a> <span class="citation">(Sheffield and Bock 2016)</span> and <a href="https://bioconductor.org/packages/regioneR" class="external-link">regioneR</a> <span class="citation">(Gel et al. 2016)</span>. Methods implemented outside
of Bioconductor include <a href="https://github.com/AndreasHeger/gat" class="external-link">GAT</a> <span class="citation">(Heger et al. 2013)</span>, <a href="https://github.com/ParkerLab/encodegsc" class="external-link">GSC</a> <span class="citation">(Bickel et al. 2010)</span>, <a href="http://bejerano.stanford.edu/great/public/html/" class="external-link">GREAT</a> <span class="citation">(McLean et al. 2010)</span>, <a href="https://github.com/favorov/GenometriCorr" class="external-link">GenometriCorr</a> <span class="citation">(Favorov et al. 2012)</span>, or <a href="http://chip-enrich.med.umich.edu/" class="external-link">ChIP-Enrich</a> <span class="citation">(Welch et al. 2014)</span>. We note that our block
bootstrapping approach closely follows that of <a href="https://github.com/ParkerLab/encodegsc" class="external-link">GSC</a>, while offering
additional features/visualizations, and is re-implemented within
R/Bioconductor with efficient vectorized code for operation on
<em>GRanges</em> objects <span class="citation">(Lawrence
2013)</span>.</p>
</div>
<div class="section level3">
<h3 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h3>
<ul>
<li>Some basic familiarity with <em>GenomicRanges</em> objects.</li>
</ul>
</div>
<div class="section level3">
<h3 id="background-reading">Background reading<a class="anchor" aria-label="anchor" href="#background-reading"></a>
</h3>
<ul>
<li><a href="https://doi.org/10.1214/10-%7BAOAS363%7D" class="external-link">Subsampling
Methods for Genomic Inference</a></li>
<li><a href="https://doi.org/10.1093/pan/mpl013" class="external-link">Matching as
Nonparametric Preprocessing for Reducing Model Dependence in Parametric
Causal Inference</a></li>
</ul>
</div>
<div class="section level3">
<h3 id="participation">Participation<a class="anchor" aria-label="anchor" href="#participation"></a>
</h3>
<p>The format is lecture + lab, with time for data exploration.</p>
</div>
<div class="section level3">
<h3 id="r-bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r-bioconductor-packages-used"></a>
</h3>
<ul>
<li>GenomicRanges</li>
<li>plyranges</li>
</ul>
</div>
<div class="section level3">
<h3 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h3>
<p>An example for a 45-minute workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Intro to nullranges</td>
<td>5m</td>
</tr>
<tr class="even">
<td>Matching with matchRanges</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Bootstrapping with bootRanges</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Q&amp;A</td>
<td>10m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h3>
</div>
<div class="section level3">
<h3 id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h3>
<ul>
<li>Understand matching vs bootstrapping for nullranges</li>
<li>Using matchRanges to generate covarite-matched subsets</li>
<li>Using bootRanges to generate block bootstrapped ranges</li>
<li>Integrating nullranges with plyranges</li>
</ul>
</div>
<div class="section level3">
<h3 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ul>
<li>matchRanges
<ul>
<li>Generate matched null sets of GenomicRanges</li>
<li>Assess quality of matching</li>
<li>Compare biological results</li>
</ul>
</li>
<li>bootRanges</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="matching-covariates-with-matchranges">Matching covariates with matchRanges<a class="anchor" aria-label="anchor" href="#matching-covariates-with-matchranges"></a>
</h2>
<p>The following content is an excerpt from the vignette <a href="https://nullranges.github.io/nullranges/articles/matching_granges.html" class="external-link">Case
study I: CTCF occupancy</a> from the nullranges package.</p>
<p>In this vignette we demonstrate generating covariate-matched,
null-hypothesis GRanges using the <code><a href="https://nullranges.github.io/nullranges/reference/matchRanges.html" class="external-link">matchRanges()</a></code> function to
test for the occupancy of CCCTC-binding factor (CTCF) at chromatin loop
anchors.</p>
<div class="section level3">
<h3 id="biological-background">Biological background<a class="anchor" aria-label="anchor" href="#biological-background"></a>
</h3>
<p>One of the fundamental principles of chromatin-looping suggests that
most loops are bound at both ends by the CTCF transcription factor (TF).
CTCF-bound loops can be formed by loop-extrusion, where the ring-like
cohesin complex extrudes chromatin until stopped by bound CTCF. By this
mechanism, we expect most loop anchors will be bound by CTCF.</p>
<p><img src="images/ctcfBoundLooping.png" title="CTCF-bound DNA Looping" alt="CTCF-bound DNA Looping" width="1147"></p>
<p>While we could test this hypothesis by simple overlap or permutation
testing, these approaches fail to account for non-uniformly distributed
covariate genomic features. For example, loop anchors are commonly bound
by CTCF and located in open chromatin regions. We can use
<code><a href="https://nullranges.github.io/nullranges/reference/matchRanges.html" class="external-link">matchRanges()</a></code> to test for CTCF occupancy at loop anchors
controlling for open chromatin regions.</p>
<p>Here, we generate a set of null-hypothesis GRanges to more rigorously
test CTCF occupancy at loop anchors independently from open chromatin
regions. We use the <code>hg19_10kb_bins</code> dataset from the
<code>nullrangesData</code> package, which contains ranges for every
10Kb bin along the genome with CTCF, DNase, and loop feature annotations
from GM12878 (see <code><a href="https://rdrr.io/pkg/nullrangesData/man/hg19_10kb_bins.html" class="external-link">?nullrangesData::hg19_10kb_bins</a></code>).</p>
<p><img src="images/MatchedGRanges.png" title="Overview of MatchGRanges" alt="Overview of MatchGRanges" width="1275"></p>
</div>
<div class="section level3">
<h3 id="matching-with-matchranges">Matching with <code>matchRanges()</code><a class="anchor" aria-label="anchor" href="#matching-with-matchranges"></a>
</h3>
<p>Before we generate our null ranges, let’s take a look at our example
dataset:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">nullrangesData</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Load example data</span></span>
<span><span class="va">bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/nullrangesData/man/hg19_10kb_bins.html" class="external-link">hg19_10kb_bins</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bins</span></span>
<span><span class="co">#&gt; GRanges object with 303641 ranges and 5 metadata columns:</span></span>
<span><span class="co">#&gt;            seqnames              ranges strand | n_ctcf_sites ctcfSignal</span></span>
<span><span class="co">#&gt;               &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;        [1]     chr1             1-10000      * |            0          0</span></span>
<span><span class="co">#&gt;        [2]     chr1         10001-20000      * |            0          0</span></span>
<span><span class="co">#&gt;        [3]     chr1         20001-30000      * |            0          0</span></span>
<span><span class="co">#&gt;        [4]     chr1         30001-40000      * |            0          0</span></span>
<span><span class="co">#&gt;        [5]     chr1         40001-50000      * |            0          0</span></span>
<span><span class="co">#&gt;        ...      ...                 ...    ... .          ...        ...</span></span>
<span><span class="co">#&gt;   [303637]     chrX 155230001-155240000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;   [303638]     chrX 155240001-155250000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;   [303639]     chrX 155250001-155260000      * |            1    4.09522</span></span>
<span><span class="co">#&gt;   [303640]     chrX 155260001-155270000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;   [303641]     chrX 155270001-155270560      * |            0    0.00000</span></span>
<span><span class="co">#&gt;            n_dnase_sites dnaseSignal    looped</span></span>
<span><span class="co">#&gt;                 &lt;factor&gt;   &lt;numeric&gt; &lt;logical&gt;</span></span>
<span><span class="co">#&gt;        [1]             0     0.00000     FALSE</span></span>
<span><span class="co">#&gt;        [2]             0     5.03572     FALSE</span></span>
<span><span class="co">#&gt;        [3]             0     0.00000     FALSE</span></span>
<span><span class="co">#&gt;        [4]             0     0.00000     FALSE</span></span>
<span><span class="co">#&gt;        [5]             0     0.00000     FALSE</span></span>
<span><span class="co">#&gt;        ...           ...         ...       ...</span></span>
<span><span class="co">#&gt;   [303637]             0     8.42068     FALSE</span></span>
<span><span class="co">#&gt;   [303638]             0     4.08961     FALSE</span></span>
<span><span class="co">#&gt;   [303639]             0     6.00443     FALSE</span></span>
<span><span class="co">#&gt;   [303640]             0     8.07179     FALSE</span></span>
<span><span class="co">#&gt;   [303641]             0     0.00000     FALSE</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 23 sequences from hg19 genome</span></span></code></pre></div>
<p><code><a href="https://nullranges.github.io/nullranges/reference/matchRanges.html" class="external-link">matchRanges()</a></code> works by selecting a set of
covariate-matched controls from a pool of options based on an input
focal set of interest. Here, we define <code>focal</code> as bins that
contain a loop anchor, <code>pool</code> as bins that don’t contain a
loop anchor, and <code>covar</code> as DNase signal and number of DNase
sites per bin:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nullranges.github.io/nullranges" class="external-link">nullranges</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Match ranges</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">mgr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://nullranges.github.io/nullranges/reference/matchRanges.html" class="external-link">matchRanges</a></span><span class="op">(</span>focal <span class="op">=</span> <span class="va">bins</span><span class="op">[</span><span class="va">bins</span><span class="op">$</span><span class="va">looped</span><span class="op">]</span>,</span>
<span>                   pool <span class="op">=</span> <span class="va">bins</span><span class="op">[</span><span class="op">!</span><span class="va">bins</span><span class="op">$</span><span class="va">looped</span><span class="op">]</span>,</span>
<span>                   covar <span class="op">=</span> <span class="op">~</span><span class="va">dnaseSignal</span> <span class="op">+</span> <span class="va">n_dnase_sites</span><span class="op">)</span></span>
<span><span class="va">mgr</span></span>
<span><span class="co">#&gt; MatchedGRanges object with 13979 ranges and 5 metadata columns:</span></span>
<span><span class="co">#&gt;           seqnames              ranges strand | n_ctcf_sites ctcfSignal</span></span>
<span><span class="co">#&gt;              &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;       [1]     chr8   16360001-16370000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;       [2]     chr1 212260001-212270000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;       [3]     chrX     9570001-9580000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;       [4]     chr9   96840001-96850000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;       [5]    chr16   31480001-31490000      * |            1    4.82341</span></span>
<span><span class="co">#&gt;       ...      ...                 ...    ... .          ...        ...</span></span>
<span><span class="co">#&gt;   [13975]    chr12 105060001-105070000      * |            1    4.29979</span></span>
<span><span class="co">#&gt;   [13976]     chr4 103530001-103540000      * |            1    5.67681</span></span>
<span><span class="co">#&gt;   [13977]    chr14   65860001-65870000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;   [13978]     chr2   96940001-96950000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;   [13979]    chr19   38910001-38920000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;           n_dnase_sites dnaseSignal    looped</span></span>
<span><span class="co">#&gt;                &lt;factor&gt;   &lt;numeric&gt; &lt;logical&gt;</span></span>
<span><span class="co">#&gt;       [1]            0      7.40552     FALSE</span></span>
<span><span class="co">#&gt;       [2]            1      9.72456     FALSE</span></span>
<span><span class="co">#&gt;       [3]            3+    10.87185     FALSE</span></span>
<span><span class="co">#&gt;       [4]            1      8.86143     FALSE</span></span>
<span><span class="co">#&gt;       [5]            3+    10.65095     FALSE</span></span>
<span><span class="co">#&gt;       ...           ...         ...       ...</span></span>
<span><span class="co">#&gt;   [13975]            3+    13.90788     FALSE</span></span>
<span><span class="co">#&gt;   [13976]            1      9.89450     FALSE</span></span>
<span><span class="co">#&gt;   [13977]            1      9.83681     FALSE</span></span>
<span><span class="co">#&gt;   [13978]            0      9.66206     FALSE</span></span>
<span><span class="co">#&gt;   [13979]            2     12.13214     FALSE</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 23 sequences from hg19 genome</span></span></code></pre></div>
<p>When the focal and pool arguments are <code>GRanges</code> objects,
<code><a href="https://nullranges.github.io/nullranges/reference/matchRanges.html" class="external-link">matchRanges()</a></code> returns a <code>MatchedGRanges</code> object.
The <code>MatchedGRanges</code> class extends <code>GRanges</code>, so
all of the same operations can be applied:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Summarize ctcfSignal by n_ctcf_sites</span></span>
<span><span class="va">mgr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">n_ctcf_sites</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>ctcfSignal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">ctcfSignal</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">n_ctcf_sites</span>, y <span class="op">=</span> <span class="va">ctcfSignal</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="nullranges_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>Here, we utilize the <a href="https://sa-lee.github.io/plyranges/" class="external-link"><code>plyranges</code>
package</a> which provides a set of “tidy” verbs for manipulating
genomic ranges for a seamless and integrated genomic analysis
workflow.</p>
</div>
<div class="section level3">
<h3 id="assessing-quality-of-matching">Assessing quality of matching<a class="anchor" aria-label="anchor" href="#assessing-quality-of-matching"></a>
</h3>
<p>We can get a quick summary of the matching quality with
<code><a href="https://nullranges.github.io/nullranges/reference/overview.html" class="external-link">overview()</a></code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://nullranges.github.io/nullranges/reference/overview.html" class="external-link">overview</a></span><span class="op">(</span><span class="va">mgr</span><span class="op">)</span></span>
<span><span class="co">#&gt; MatchedGRanges object: </span></span>
<span><span class="co">#&gt;        set      N dnaseSignal.mean dnaseSignal.sd n_dnase_sites.0</span></span>
<span><span class="co">#&gt;      focal  13979             10.0            1.9            2341</span></span>
<span><span class="co">#&gt;    matched  13979             10.0            1.9            2297</span></span>
<span><span class="co">#&gt;       pool 289662              7.9            2.7          222164</span></span>
<span><span class="co">#&gt;  unmatched 275683              7.8            2.7          219867</span></span>
<span><span class="co">#&gt;  n_dnase_sites.1 n_dnase_sites.2 n_dnase_sites.3+ ps.mean ps.sd</span></span>
<span><span class="co">#&gt;             4829            2353             4456   0.130 0.072</span></span>
<span><span class="co">#&gt;             5187            2487             4008   0.130 0.071</span></span>
<span><span class="co">#&gt;            34826           13627            19045   0.042 0.061</span></span>
<span><span class="co">#&gt;            29639           11140            15037   0.037 0.057</span></span>
<span><span class="co">#&gt; --------</span></span>
<span><span class="co">#&gt; focal - matched: </span></span>
<span><span class="co">#&gt;  dnaseSignal.mean dnaseSignal.sd n_dnase_sites.0 n_dnase_sites.1</span></span>
<span><span class="co">#&gt;             0.037        -0.0061              44            -360</span></span>
<span><span class="co">#&gt;  n_dnase_sites.2 n_dnase_sites.3+ ps.mean   ps.sd</span></span>
<span><span class="co">#&gt;             -130              450 0.00025 0.00081</span></span></code></pre></div>
<p>For continuous covariates (such as <code>dnaseSignal</code>),
<code><a href="https://nullranges.github.io/nullranges/reference/overview.html" class="external-link">overview()</a></code> shows the mean and standard deviation between
each matched set. For categorical covariates, such as
<code>n_dnase_sites</code>, <code><a href="https://nullranges.github.io/nullranges/reference/overview.html" class="external-link">overview()</a></code> reports the number
of observations per category and matched set. The bottom section shows
the mean and s.d (or n, for factors) difference between focal and
matched sets.</p>
<p><code><a href="https://nullranges.github.io/nullranges/reference/overview.html" class="external-link">overview()</a></code> also summarizes the propensity scores for
each set to give a quick idea of overall matching quality.</p>
<div class="section level4">
<h4 id="visualizing-matching-results">Visualizing matching results<a class="anchor" aria-label="anchor" href="#visualizing-matching-results"></a>
</h4>
<p>Let’s visualize overall matching quality by plotting propensity
scores for the focal, pool, and matched sets:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://nullranges.github.io/nullranges/reference/plotPropensity.html" class="external-link">plotPropensity</a></span><span class="op">(</span><span class="va">mgr</span>, sets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'f'</span>, <span class="st">'p'</span>, <span class="st">'m'</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">'ridges'</span><span class="op">)</span></span></code></pre></div>
<p><img src="nullranges_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>From this plot, it is clear that the matched set is much closer to
the focal set than the pool set.</p>
<p>We can ensure that covariate distributions have been matched
appropriately by using the <code><a href="https://nullranges.github.io/nullranges/reference/covariates.html" class="external-link">covariates()</a></code> function to extract
matched covariates along with <code>patchwork</code> and
<code>plotCovarite</code> to visualize all distributions:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://nullranges.github.io/nullranges/reference/covariates.html" class="external-link">covariates</a></span><span class="op">(</span><span class="va">mgr</span><span class="op">)</span>, <span class="va">plotCovariate</span>, x<span class="op">=</span><span class="va">mgr</span>, sets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'f'</span>, <span class="st">'m'</span>, <span class="st">'p'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="st">'/'</span>, <span class="va">plots</span><span class="op">)</span></span></code></pre></div>
<p><img src="nullranges_files/figure-html/unnamed-chunk-9-1.png" width="480"></p>
</div>
</div>
<div class="section level3">
<h3 id="compare-ctcf-sites">Compare CTCF sites<a class="anchor" aria-label="anchor" href="#compare-ctcf-sites"></a>
</h3>
<p>Using our matched ranges, we can compare CTCF occupancy in bins that
1) contain a loop anchor (i.e. looped), 2) don’t contain a loop anchor
(i.e. unlooped), or 3) don’t contain a loop anchor, but are also matched
for the strength and number of DNase sites (i.e. matched). In this case,
we calculate CTCF occupancy as the percent of bins that contain CTCF
among our 3 sets by using the <code><a href="https://nullranges.github.io/nullranges/reference/focal.html" class="external-link">focal()</a></code> and
<code><a href="https://nullranges.github.io/nullranges/reference/pool.html" class="external-link">pool()</a></code> accessor functions:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Percent of bins with CTCF</span></span>
<span><span class="va">g1</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://nullranges.github.io/nullranges/reference/focal.html" class="external-link">focal</a></span><span class="op">(</span><span class="va">mgr</span><span class="op">)</span><span class="op">$</span><span class="va">n_ctcf_sites</span> <span class="op">&gt;=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://nullranges.github.io/nullranges/reference/focal.html" class="external-link">focal</a></span><span class="op">(</span><span class="va">mgr</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span>
<span><span class="va">g2</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://nullranges.github.io/nullranges/reference/pool.html" class="external-link">pool</a></span><span class="op">(</span><span class="va">mgr</span><span class="op">)</span><span class="op">$</span><span class="va">n_ctcf_sites</span> <span class="op">&gt;=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://nullranges.github.io/nullranges/reference/pool.html" class="external-link">pool</a></span><span class="op">(</span><span class="va">mgr</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span>
<span><span class="va">g3</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">mgr</span><span class="op">$</span><span class="va">n_ctcf_sites</span> <span class="op">&gt;=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">mgr</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span>
<span></span>
<span><span class="co">## Visualize</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span>height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">g1</span>, <span class="va">g2</span>, <span class="va">g3</span><span class="op">)</span>,</span>
<span>        names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'looped\n(focal)'</span>, <span class="st">'unlooped\n(pool)'</span>, <span class="st">'unlooped\n(matched)'</span><span class="op">)</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"CTCF occupied bins (%)"</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'#1F78B4'</span>, <span class="st">'#33A02C'</span>, <span class="st">'#A6CEE3'</span><span class="op">)</span>,</span>
<span>        main <span class="op">=</span> <span class="st">'CTCF occupancy'</span>,</span>
<span>        border <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>        las <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="nullranges_files/figure-html/unnamed-chunk-10-1.png" width="432"></p>
</div>
</div>
<div class="section level2">
<h2 id="block-bootstrapping-with-bootranges">Block bootstrapping with bootRanges<a class="anchor" aria-label="anchor" href="#block-bootstrapping-with-bootranges"></a>
</h2>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="additional-references">Additional references<a class="anchor" aria-label="anchor" href="#additional-references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-bickel_2010" class="csl-entry">
Bickel, Peter J., Nathan Boley, James B. Brown, Haiyan Huang, and Nancy
R. Zhang. 2010. <span>“Subsampling Methods for Genomic
Inference.”</span> <em>The Annals of Applied Statistics</em> 4 (4):
1660–97. <a href="https://doi.org/10.1214/10-%7BAOAS363%7D" class="external-link">https://doi.org/10.1214/10-{AOAS363}</a>.
</div>
<div id="ref-de_2014" class="csl-entry">
De, Subhajyoti, Brent S Pedersen, and Katerina Kechris. 2014. <span>“The
Dilemma of Choosing the Ideal Permutation Strategy While Estimating
Statistical Significance of Genome-Wide Enrichment.”</span>
<em>Briefings in Bioinformatics</em> 15 (6): 919–28. <a href="https://doi.org/10.1093/bib/bbt053" class="external-link">https://doi.org/10.1093/bib/bbt053</a>.
</div>
<div id="ref-GenometriCorr" class="csl-entry">
Favorov, Alexander, Loris Mularoni, Leslie M Cope, Yulia Medvedeva,
Andrey A Mironov, Vsevolod J Makeev, and Sarah J Wheelan. 2012.
<span>“Exploring Massive, Genome Scale Datasets with the
<span>GenometriCorr</span> Package.”</span> <em><span>PLoS</span>
Computational Biology</em> 8 (5): e1002529. <a href="https://doi.org/10.1371/journal.pcbi.1002529" class="external-link">https://doi.org/10.1371/journal.pcbi.1002529</a>.
</div>
<div id="ref-regioneR" class="csl-entry">
Gel, Bernat, Anna Díez-Villanueva, Eduard Serra, Marcus Buschbeck,
Miguel A Peinado, and Roberto Malinverni. 2016. <span>“<span class="nocase">regioneR</span>: An r/Bioconductor Package for the
Association Analysis of Genomic Regions Based on Permutation
Tests.”</span> <em>Bioinformatics</em> 32 (2): 289–91. <a href="https://doi.org/10.1093/bioinformatics/btv562" class="external-link">https://doi.org/10.1093/bioinformatics/btv562</a>.
</div>
<div id="ref-haiminen_2007" class="csl-entry">
Haiminen, Niina, Heikki Mannila, and Evimaria Terzi. 2007.
<span>“Comparing Segmentations by Applying Randomization
Techniques.”</span> <em><span>BMC</span> Bioinformatics</em> 8 (May):
171. <a href="https://doi.org/10.1186/1471-2105-8-171" class="external-link">https://doi.org/10.1186/1471-2105-8-171</a>.
</div>
<div id="ref-GAT" class="csl-entry">
Heger, Andreas, Caleb Webber, Martin Goodson, Chris P Ponting, and
Gerton Lunter. 2013. <span>“<span>GAT</span>: A Simulation Framework for
Testing the Association of Genomic Intervals.”</span>
<em>Bioinformatics</em> 29 (16): 2046–48. <a href="https://doi.org/10.1093/bioinformatics/btt343" class="external-link">https://doi.org/10.1093/bioinformatics/btt343</a>.
</div>
<div id="ref-huen_2010" class="csl-entry">
Huen, David S, and Steven Russell. 2010. <span>“On the Use of Resampling
Tests for Evaluating Statistical Significance of Binding-Site
Co-Occurrence.”</span> <em><span>BMC</span> Bioinformatics</em> 11
(June): 359. <a href="https://doi.org/10.1186/1471-2105-11-359" class="external-link">https://doi.org/10.1186/1471-2105-11-359</a>.
</div>
<div id="ref-kanduri_2019" class="csl-entry">
Kanduri, Chakravarthi, Christoph Bock, Sveinung Gundersen, Eivind Hovig,
and Geir Kjetil Sandve. 2019. <span>“Colocalization Analyses of Genomic
Elements: Approaches, Recommendations and Challenges.”</span>
<em>Bioinformatics</em> 35 (9): 1615–24. <a href="https://doi.org/10.1093/bioinformatics/bty835" class="external-link">https://doi.org/10.1093/bioinformatics/bty835</a>.
</div>
<div id="ref-granges" class="csl-entry">
Lawrence, Wolfgang AND Pagès, Michael AND Huber. 2013. <span>“Software
for Computing and Annotating Genomic Ranges.”</span> <em>PLOS
Computational Biology</em> 9 (8): 1–10. <a href="https://doi.org/10.1371/journal.pcbi.1003118" class="external-link">https://doi.org/10.1371/journal.pcbi.1003118</a>.
</div>
<div id="ref-GREAT" class="csl-entry">
McLean, Cory Y., Dave Bristor, Michael Hiller, Shoa L. Clarke, Bruce T.
Schaar, Craig B. Lowe, Aaron M. Wenger, and Gill Bejerano. 2010.
<span>“GREAT Improves Functional Interpretation of Cis-Regulatory
Regions.”</span> <em>Nature Biotechnology</em> 28 (5): 495–501. <a href="https://doi.org/10.1038/nbt.1630" class="external-link">https://doi.org/10.1038/nbt.1630</a>.
</div>
<div id="ref-LOLA" class="csl-entry">
Sheffield, Nathan C, and Christoph Bock. 2016. <span>“<span>LOLA</span>:
Enrichment Analysis for Genomic Region Sets and Regulatory Elements in r
and Bioconductor.”</span> <em>Bioinformatics</em> 32 (4): 587–89. <a href="https://doi.org/10.1093/bioinformatics/btv612" class="external-link">https://doi.org/10.1093/bioinformatics/btv612</a>.
</div>
<div id="ref-ChIP-Enrich" class="csl-entry">
Welch, Ryan P., Chee Lee, Paul M. Imbriano, Snehal Patil, Terry E.
Weymouth, R. Alex Smith, Laura J. Scott, and Maureen A. Sartor. 2014.
<span>“<span class="nocase">ChIP-Enrich: gene set enrichment testing for
ChIP-seq data</span>.”</span> <em>Nucleic Acids Research</em> 42 (13):
e105–5. <a href="https://doi.org/10.1093/nar/gku463" class="external-link">https://doi.org/10.1093/nar/gku463</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wancen Mu, Eric Davis, Mikhail Dozmorov, Stuart Lee, Michael Love, Douglas Phanstiel.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
