<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>nullranges: Generation of nullranges via bootstrapping or covariate matching • Bioc2022nullranges</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="nullranges: Generation of nullranges via bootstrapping or covariate matching">
<meta property="og:description" content="Bioc2022nullranges">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Bioc2022nullranges</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/nullranges.html">nullranges: Generation of nullranges via bootstrapping or covariate matching</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/EricSDavis/Bioc2022nullranges" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>nullranges: Generation of nullranges via
bootstrapping or covariate matching</h1>
                        <h4 data-toc-skip class="author">Wancen Mu, Eric
S. Davis</h4>
            
            <h4 data-toc-skip class="date">July 2022</h4>
      
      
      <div class="hidden name"><code>nullranges.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h3>
<p>The nullranges package contains functions for generation of feature
sets (genomic regions) for exploring the null hypothesis of overlap or
colocalization of two observed feature sets. The package has two
approaches for generating null feature sets, <em>matching</em> and
<em>bootstrapping</em>:</p>
<ul>
<li>
<p>Matched subsampling with matchRanges:</p>
<p>Subsampling from a pool of features, but controlling for certain
characteristics.</p>
</li>
<li>
<p>Bootstrapping with bootRanges:</p>
<p>Subsampling from original features, but controlling for their spatial
pattern</p>
</li>
</ul>
<p>In this workshop, we will demonstrate the two major approaches for
generating null hypothesis genomic ranges with the
<code>nullranges</code> package.</p>
</div>
<div class="section level3">
<h3 id="background-and-other-methods">Background and other methods<a class="anchor" aria-label="anchor" href="#background-and-other-methods"></a>
</h3>
<p>Suppose we want to examine the significance of overlaps of genomic
sets of features <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span>. To test the significance of this
overlap, we calculate the overlap expected under the null by generating
a null feature set <span class="math inline">\(y'\)</span>
(potentially many times). The null features in <span class="math inline">\(y'\)</span> may be characterized by:</p>
<ol style="list-style-type: decimal">
<li>Drawing from a larger pool <span class="math inline">\(z\)</span>
(<span class="math inline">\(y' \subset z\)</span>), such that <span class="math inline">\(y\)</span> and <span class="math inline">\(y'\)</span> have a similar distribution over
one or more covariates. This is the “matching” case. Note that the
features in <span class="math inline">\(y'\)</span> are original
features, just drawn from a different pool than y.</li>
<li>Generating a new set of genomic features <span class="math inline">\(y'\)</span>, constructing them from the
original set <span class="math inline">\(y\)</span> by selecting blocks
of the genome with replacement, i.e. such that features can be sampled
more than once. This is the “bootstrapping” case. Note that, in this
case, <span class="math inline">\(y'\)</span> is an artificial
feature set, although the re-sampled features can retain covariates such
as score from the original feature set <span class="math inline">\(y\)</span>.</li>
</ol>
<div class="section level4">
<h4 id="in-other-words">In other words<a class="anchor" aria-label="anchor" href="#in-other-words"></a>
</h4>
<ol style="list-style-type: decimal">
<li>Matching – drawing from a pool of features but controlling for
certain characteristics</li>
<li>Block Bootstrapping – placing a number of artificial features in the
genome but controlling for their spatial pattern</li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="releated-work">Releated work<a class="anchor" aria-label="anchor" href="#releated-work"></a>
</h3>
<p>For general considerations of generation of null feature sets or
segmentation for enrichment or colocalization analysis, consider the
papers of <span class="citation">De, Pedersen, and Kechris
(2014)</span>, <span class="citation">Haiminen, Mannila, and Terzi
(2007)</span>, <span class="citation">Huen and Russell (2010)</span>,
and <span class="citation">Kanduri et al. (2019)</span> (with links in
references below). Other Bioconductor packages that offer randomization
techniques for enrichment analysis include: <a href="https://cran.r-project.org/web/packages/MatchIt/vignettes/MatchIt.html" class="external-link">MatchIt</a>
<a href="https://bioconductor.org/packages/LOLA" class="external-link">LOLA</a> <span class="citation">(Sheffield and Bock 2016)</span> and <a href="https://bioconductor.org/packages/regioneR" class="external-link">regioneR</a> <span class="citation">(Gel et al. 2016)</span>. Methods implemented outside
of Bioconductor include <a href="https://github.com/AndreasHeger/gat" class="external-link">GAT</a> <span class="citation">(Heger et al. 2013)</span>, <a href="https://github.com/ParkerLab/encodegsc" class="external-link">GSC</a> <span class="citation">(Bickel et al. 2010)</span>, <a href="http://bejerano.stanford.edu/great/public/html/" class="external-link">GREAT</a> <span class="citation">(McLean et al. 2010)</span>, <a href="https://github.com/favorov/GenometriCorr" class="external-link">GenometriCorr</a> <span class="citation">(Favorov et al. 2012)</span>, or <a href="http://chip-enrich.med.umich.edu/" class="external-link">ChIP-Enrich</a> <span class="citation">(Welch et al. 2014)</span>. Note that our block
bootstrapping idea motivated from <a href="https://github.com/ParkerLab/encodegsc" class="external-link">GSC</a>. In order to
address computationally intensive issue, <a href="https://github.com/ParkerLab/encodegsc" class="external-link">GSC</a> implements a
strategy of swapping pairs of blocks in order to generate bootstrap
distribution while avoiding generation of a genome-scale bootstrap
sample. We uses efficient vectorized code for operation on
<em>GRanges</em> objects <span class="citation">(Lawrence 2013)</span>
to offer <span class="math inline">\(\color{brown}{\text{genome-scale}}\)</span>
bootstrap data with additional features/visualizations, and is
re-implemented within R/Bioconductor.</p>
</div>
<div class="section level3">
<h3 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h3>
<ul>
<li>Some basic familiarity with <em>GenomicRanges</em> objects.</li>
</ul>
</div>
<div class="section level3">
<h3 id="background-reading">Background reading<a class="anchor" aria-label="anchor" href="#background-reading"></a>
</h3>
<ul>
<li><a href="https://www.jstor.org/stable/23362444" class="external-link">Subsampling Methods
for Genomic Inference</a></li>
<li><a href="https://doi.org/10.1093/pan/mpl013" class="external-link">Matching as
Nonparametric Preprocessing for Reducing Model Dependence in Parametric
Causal Inference</a></li>
</ul>
</div>
<div class="section level3">
<h3 id="participation">Participation<a class="anchor" aria-label="anchor" href="#participation"></a>
</h3>
<p>The format is lecture + lab, with time for data exploration.</p>
</div>
<div class="section level3">
<h3 id="r-bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r-bioconductor-packages-used"></a>
</h3>
<ul>
<li>GenomicRanges</li>
<li>plyranges</li>
</ul>
</div>
<div class="section level3">
<h3 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h3>
<p>An example for a 45-minute workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Intro to nullranges</td>
<td>5m</td>
</tr>
<tr class="even">
<td>Matching with matchRanges</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Bootstrapping with bootRanges</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Q&amp;A</td>
<td>10m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h3>
</div>
<div class="section level3">
<h3 id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h3>
<ul>
<li>Understand matching vs bootstrapping for nullranges</li>
<li>Using matchRanges to generate covarite-matched subsets</li>
<li>Using bootRanges to generate block bootstrapped ranges</li>
<li>Integrating nullranges with plyranges</li>
</ul>
</div>
<div class="section level3">
<h3 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ul>
<li>matchRanges
<ul>
<li>Generate matched null sets of GenomicRanges</li>
<li>Assess quality of matching</li>
<li>Compare biological results</li>
</ul>
</li>
<li>bootRanges
<ul>
<li>Generate block bootstrapped null sets of GenomicRanges</li>
<li>Assess quality of subsampling sets</li>
<li>Use with plyranges to derive biological results</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="matching-covariates-with-matchranges">Matching covariates with matchRanges<a class="anchor" aria-label="anchor" href="#matching-covariates-with-matchranges"></a>
</h2>
<div class="section level3">
<h3 id="overview-1">Overview<a class="anchor" aria-label="anchor" href="#overview-1"></a>
</h3>
<p><code><a href="https://nullranges.github.io/nullranges/reference/matchRanges.html" class="external-link">matchRanges()</a></code> allows users to subsample a pool of ranges
such that the resulting matched set contains similar distributions of
covariates (i.e., genomic features) as a focal set of interest.</p>
<p><img src="images/MatchedGRanges.png" title="Overview of MatchGRanges" alt="Overview of MatchGRanges" width="1275"></p>
<p>Resulting sets can then be compared, without potential confounding
effects from covariates.</p>
</div>
<div class="section level3">
<h3 id="biological-background">Biological background<a class="anchor" aria-label="anchor" href="#biological-background"></a>
</h3>
<p>Most chromatin loops are formed in a processes called loop extrusion,
where the ring-like cohesin complex extrudes chromatin until stopped by
bound CTCF transcription factors (TFs). Therefore, most chromatin loops
tend to be bound at both ends by CTCF. However, the anchors of chromatin
loops are also located in accessible chromatin regions and act as
potential confounders.</p>
<p><img src="images/ctcfBoundLooping.png" title="CTCF-bound DNA Looping" alt="CTCF-bound DNA Looping" width="50%" style="display: block; margin: auto;"></p>
<p>Suppose we wanted to compare CTCF occupancy between the anchors of
looped ranges and unlooped ranges. <code><a href="https://nullranges.github.io/nullranges/reference/matchRanges.html" class="external-link">matchRanges()</a></code> can help
by generating a null set of ranges controlling for potential confounding
by chromatin accessibility.</p>
<p>The <code>hg19_10kb_bins</code> dataset from the
<code>nullrangesData</code> package contains ranges for every 10Kb bin
along the genome with CTCF, DNase, and loop feature annotations
fromGM12878 (see <code><a href="https://rdrr.io/pkg/nullrangesData/man/hg19_10kb_bins.html" class="external-link">?nullrangesData::hg19_10kb_bins</a></code>).</p>
</div>
<div class="section level3">
<h3 id="matching-with-matchranges">Matching with <code>matchRanges()</code><a class="anchor" aria-label="anchor" href="#matching-with-matchranges"></a>
</h3>
<p>Before we generate our null ranges, let’s take a look at our example
dataset:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">nullrangesData</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Load example data</span></span>
<span><span class="va">bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/nullrangesData/man/hg19_10kb_bins.html" class="external-link">hg19_10kb_bins</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bins</span></span>
<span><span class="co">#&gt; GRanges object with 303641 ranges and 5 metadata columns:</span></span>
<span><span class="co">#&gt;            seqnames              ranges strand | n_ctcf_sites ctcfSignal</span></span>
<span><span class="co">#&gt;               &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;        [1]     chr1             1-10000      * |            0          0</span></span>
<span><span class="co">#&gt;        [2]     chr1         10001-20000      * |            0          0</span></span>
<span><span class="co">#&gt;        [3]     chr1         20001-30000      * |            0          0</span></span>
<span><span class="co">#&gt;        [4]     chr1         30001-40000      * |            0          0</span></span>
<span><span class="co">#&gt;        [5]     chr1         40001-50000      * |            0          0</span></span>
<span><span class="co">#&gt;        ...      ...                 ...    ... .          ...        ...</span></span>
<span><span class="co">#&gt;   [303637]     chrX 155230001-155240000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;   [303638]     chrX 155240001-155250000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;   [303639]     chrX 155250001-155260000      * |            1    4.09522</span></span>
<span><span class="co">#&gt;   [303640]     chrX 155260001-155270000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;   [303641]     chrX 155270001-155270560      * |            0    0.00000</span></span>
<span><span class="co">#&gt;            n_dnase_sites dnaseSignal    looped</span></span>
<span><span class="co">#&gt;                 &lt;factor&gt;   &lt;numeric&gt; &lt;logical&gt;</span></span>
<span><span class="co">#&gt;        [1]             0     0.00000     FALSE</span></span>
<span><span class="co">#&gt;        [2]             0     5.03572     FALSE</span></span>
<span><span class="co">#&gt;        [3]             0     0.00000     FALSE</span></span>
<span><span class="co">#&gt;        [4]             0     0.00000     FALSE</span></span>
<span><span class="co">#&gt;        [5]             0     0.00000     FALSE</span></span>
<span><span class="co">#&gt;        ...           ...         ...       ...</span></span>
<span><span class="co">#&gt;   [303637]             0     8.42068     FALSE</span></span>
<span><span class="co">#&gt;   [303638]             0     4.08961     FALSE</span></span>
<span><span class="co">#&gt;   [303639]             0     6.00443     FALSE</span></span>
<span><span class="co">#&gt;   [303640]             0     8.07179     FALSE</span></span>
<span><span class="co">#&gt;   [303641]             0     0.00000     FALSE</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 23 sequences from hg19 genome</span></span></code></pre></div>
<p><code><a href="https://nullranges.github.io/nullranges/reference/matchRanges.html" class="external-link">matchRanges()</a></code> works by selecting a set of
covariate-matched controls from a pool of options based on an input
focal set of interest. Here, we define <code>focal</code> as bins that
contain a loop anchor, <code>pool</code> as bins that don’t contain a
loop anchor, and <code>covar</code> as DNase signal and number of DNase
sites per bin:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nullranges.github.io/nullranges" class="external-link">nullranges</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Match ranges</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">mgr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://nullranges.github.io/nullranges/reference/matchRanges.html" class="external-link">matchRanges</a></span><span class="op">(</span>focal <span class="op">=</span> <span class="va">bins</span><span class="op">[</span><span class="va">bins</span><span class="op">$</span><span class="va">looped</span><span class="op">]</span>,</span>
<span>                   pool <span class="op">=</span> <span class="va">bins</span><span class="op">[</span><span class="op">!</span><span class="va">bins</span><span class="op">$</span><span class="va">looped</span><span class="op">]</span>,</span>
<span>                   covar <span class="op">=</span> <span class="op">~</span><span class="va">dnaseSignal</span> <span class="op">+</span> <span class="va">n_dnase_sites</span><span class="op">)</span></span>
<span><span class="va">mgr</span></span>
<span><span class="co">#&gt; MatchedGRanges object with 13979 ranges and 5 metadata columns:</span></span>
<span><span class="co">#&gt;           seqnames              ranges strand | n_ctcf_sites ctcfSignal</span></span>
<span><span class="co">#&gt;              &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;       [1]     chr8   16360001-16370000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;       [2]     chr1 212260001-212270000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;       [3]     chrX     9570001-9580000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;       [4]     chr9   96840001-96850000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;       [5]    chr16   31480001-31490000      * |            1    4.82341</span></span>
<span><span class="co">#&gt;       ...      ...                 ...    ... .          ...        ...</span></span>
<span><span class="co">#&gt;   [13975]    chr12 105060001-105070000      * |            1    4.29979</span></span>
<span><span class="co">#&gt;   [13976]     chr4 103530001-103540000      * |            1    5.67681</span></span>
<span><span class="co">#&gt;   [13977]    chr14   65860001-65870000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;   [13978]     chr2   96940001-96950000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;   [13979]    chr19   38910001-38920000      * |            0    0.00000</span></span>
<span><span class="co">#&gt;           n_dnase_sites dnaseSignal    looped</span></span>
<span><span class="co">#&gt;                &lt;factor&gt;   &lt;numeric&gt; &lt;logical&gt;</span></span>
<span><span class="co">#&gt;       [1]            0      7.40552     FALSE</span></span>
<span><span class="co">#&gt;       [2]            1      9.72456     FALSE</span></span>
<span><span class="co">#&gt;       [3]            3+    10.87185     FALSE</span></span>
<span><span class="co">#&gt;       [4]            1      8.86143     FALSE</span></span>
<span><span class="co">#&gt;       [5]            3+    10.65095     FALSE</span></span>
<span><span class="co">#&gt;       ...           ...         ...       ...</span></span>
<span><span class="co">#&gt;   [13975]            3+    13.90788     FALSE</span></span>
<span><span class="co">#&gt;   [13976]            1      9.89450     FALSE</span></span>
<span><span class="co">#&gt;   [13977]            1      9.83681     FALSE</span></span>
<span><span class="co">#&gt;   [13978]            0      9.66206     FALSE</span></span>
<span><span class="co">#&gt;   [13979]            2     12.13214     FALSE</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 23 sequences from hg19 genome</span></span></code></pre></div>
<p>When the focal and pool arguments are <code>GRanges</code> objects,
<code><a href="https://nullranges.github.io/nullranges/reference/matchRanges.html" class="external-link">matchRanges()</a></code> returns a <code>MatchedGRanges</code> object.
The <code>MatchedGRanges</code> class extends <code>GRanges</code>, so
all of the same operations can be applied:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Summarize ctcfSignal by n_ctcf_sites</span></span>
<span><span class="va">mgr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">n_ctcf_sites</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>ctcfSignal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">ctcfSignal</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">n_ctcf_sites</span>, y <span class="op">=</span> <span class="va">ctcfSignal</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="nullranges_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>Here, we utilize the <a href="https://sa-lee.github.io/plyranges/" class="external-link"><code>plyranges</code>
package</a> which provides a set of “tidy” verbs for manipulating
genomic ranges for a seamless and integrated genomic analysis
workflow.</p>
</div>
<div class="section level3">
<h3 id="assessing-quality-of-matching">Assessing quality of matching<a class="anchor" aria-label="anchor" href="#assessing-quality-of-matching"></a>
</h3>
<p>We can get a quick summary of the matching quality with
<code><a href="https://nullranges.github.io/nullranges/reference/overview.html" class="external-link">overview()</a></code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://nullranges.github.io/nullranges/reference/overview.html" class="external-link">overview</a></span><span class="op">(</span><span class="va">mgr</span><span class="op">)</span></span>
<span><span class="co">#&gt; MatchedGRanges object: </span></span>
<span><span class="co">#&gt;        set      N dnaseSignal.mean dnaseSignal.sd n_dnase_sites.0</span></span>
<span><span class="co">#&gt;      focal  13979             10.0            1.9            2341</span></span>
<span><span class="co">#&gt;    matched  13979             10.0            1.9            2297</span></span>
<span><span class="co">#&gt;       pool 289662              7.9            2.7          222164</span></span>
<span><span class="co">#&gt;  unmatched 275683              7.8            2.7          219867</span></span>
<span><span class="co">#&gt;  n_dnase_sites.1 n_dnase_sites.2 n_dnase_sites.3+ ps.mean ps.sd</span></span>
<span><span class="co">#&gt;             4829            2353             4456   0.130 0.072</span></span>
<span><span class="co">#&gt;             5187            2487             4008   0.130 0.071</span></span>
<span><span class="co">#&gt;            34826           13627            19045   0.042 0.061</span></span>
<span><span class="co">#&gt;            29639           11140            15037   0.037 0.057</span></span>
<span><span class="co">#&gt; --------</span></span>
<span><span class="co">#&gt; focal - matched: </span></span>
<span><span class="co">#&gt;  dnaseSignal.mean dnaseSignal.sd n_dnase_sites.0 n_dnase_sites.1</span></span>
<span><span class="co">#&gt;             0.037        -0.0061              44            -360</span></span>
<span><span class="co">#&gt;  n_dnase_sites.2 n_dnase_sites.3+ ps.mean   ps.sd</span></span>
<span><span class="co">#&gt;             -130              450 0.00025 0.00081</span></span></code></pre></div>
<p>For continuous covariates (such as <code>dnaseSignal</code>),
<code><a href="https://nullranges.github.io/nullranges/reference/overview.html" class="external-link">overview()</a></code> shows the mean and standard deviation between
each matched set. For categorical covariates, such as
<code>n_dnase_sites</code>, <code><a href="https://nullranges.github.io/nullranges/reference/overview.html" class="external-link">overview()</a></code> reports the number
of observations per category and matched set. The bottom section shows
the mean and s.d (or n, for factors) difference between focal and
matched sets.</p>
<div class="section level4">
<h4 id="visualizing-matching-results">Visualizing matching results<a class="anchor" aria-label="anchor" href="#visualizing-matching-results"></a>
</h4>
<p>Let’s visualize covariate balance by plotting propensity scores for
the focal, pool, and matched sets:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://nullranges.github.io/nullranges/reference/plotPropensity.html" class="external-link">plotPropensity</a></span><span class="op">(</span><span class="va">mgr</span>, sets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'f'</span>, <span class="st">'p'</span>, <span class="st">'m'</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">'ridges'</span><span class="op">)</span></span></code></pre></div>
<p><img src="nullranges_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>These plots show that the distributions of covariates in the matched
set are similar to the focal set.</p>
<p>We can ensure that covariate distributions have been matched
appropriately by using the <code><a href="https://nullranges.github.io/nullranges/reference/covariates.html" class="external-link">covariates()</a></code> function to extract
matched covariates along with <code>patchwork</code> and
<code>plotCovariate</code> to visualize all distributions:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://nullranges.github.io/nullranges/reference/covariates.html" class="external-link">covariates</a></span><span class="op">(</span><span class="va">mgr</span><span class="op">)</span>, <span class="va">plotCovariate</span>, x<span class="op">=</span><span class="va">mgr</span>, sets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'f'</span>, <span class="st">'m'</span>, <span class="st">'p'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="st">'/'</span>, <span class="va">plots</span><span class="op">)</span></span></code></pre></div>
<p><img src="nullranges_files/figure-html/unnamed-chunk-9-1.png" width="480"></p>
</div>
</div>
<div class="section level3">
<h3 id="compare-ctcf-sites">Compare CTCF sites<a class="anchor" aria-label="anchor" href="#compare-ctcf-sites"></a>
</h3>
<p>Using our matched ranges, we can compare CTCF occupancy in bins that
1) contain a loop anchor (i.e. looped), 2) don’t contain a loop anchor
(i.e. unlooped), or 3) don’t contain a loop anchor, but are also matched
for the strength and number of DNase sites (i.e. matched). In this case,
we calculate CTCF occupancy as the percent of bins that contain CTCF
among our 3 sets by using the <code><a href="https://nullranges.github.io/nullranges/reference/focal.html" class="external-link">focal()</a></code> and
<code><a href="https://nullranges.github.io/nullranges/reference/pool.html" class="external-link">pool()</a></code> accessor functions:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Percent of bins with CTCF</span></span>
<span><span class="va">g1</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://nullranges.github.io/nullranges/reference/focal.html" class="external-link">focal</a></span><span class="op">(</span><span class="va">mgr</span><span class="op">)</span><span class="op">$</span><span class="va">n_ctcf_sites</span> <span class="op">&gt;=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://nullranges.github.io/nullranges/reference/focal.html" class="external-link">focal</a></span><span class="op">(</span><span class="va">mgr</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span>
<span><span class="va">g2</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://nullranges.github.io/nullranges/reference/pool.html" class="external-link">pool</a></span><span class="op">(</span><span class="va">mgr</span><span class="op">)</span><span class="op">$</span><span class="va">n_ctcf_sites</span> <span class="op">&gt;=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://nullranges.github.io/nullranges/reference/pool.html" class="external-link">pool</a></span><span class="op">(</span><span class="va">mgr</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span>
<span><span class="va">g3</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">mgr</span><span class="op">$</span><span class="va">n_ctcf_sites</span> <span class="op">&gt;=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">mgr</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span>
<span></span>
<span><span class="co">## Visualize</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span>height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">g1</span>, <span class="va">g2</span>, <span class="va">g3</span><span class="op">)</span>,</span>
<span>        names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'looped\n(focal)'</span>, <span class="st">'unlooped\n(pool)'</span>, <span class="st">'unlooped\n(matched)'</span><span class="op">)</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"CTCF occupied bins (%)"</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'#1F78B4'</span>, <span class="st">'#33A02C'</span>, <span class="st">'#A6CEE3'</span><span class="op">)</span>,</span>
<span>        main <span class="op">=</span> <span class="st">'CTCF occupancy'</span>,</span>
<span>        border <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>        las <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="nullranges_files/figure-html/unnamed-chunk-10-1.png" width="432"></p>
</div>
</div>
<div class="section level2">
<h2 id="block-bootstrapping-with-bootranges">Block bootstrapping with bootRanges<a class="anchor" aria-label="anchor" href="#block-bootstrapping-with-bootranges"></a>
</h2>
<div class="section level3">
<h3 id="biological-background-1">Biological background<a class="anchor" aria-label="anchor" href="#biological-background-1"></a>
</h3>
<p>One strategy for generating a null distribution for a dataset that
doesn’t have a focal/pool structure as above is to permute or shuffle
the genomic features, possibly while considering an exclusion list of
regions where features should not be located. However, genomic feature
sets often exhibit a <span class="math inline">\(\color{olive}{\text{complex dependency
structure}}\)</span>, both in terms of their <span class="math inline">\(\color{brown}{\text{spatial pattern}}\)</span>,
and in terms of <span class="math inline">\(\color{brown}{\text{local
correlation of metadata}}\)</span> (signal strength,
cell-type-specificity, etc.) Naive permutation ignores the dependence
between positions and thus won’t exhibits natural clumping properties.
An alternative way is to generate random feature sets by sampling large
blocks of features within the segments from the original set with
replacement, as proposed by <span class="citation">Bickel et al.
(2010)</span>.</p>
<p>Below, we see that various features (GC content, gene density, and
CRE location) exhibit spatial patterns easily discernible at the scale
of 500kb.</p>
<p><img src="images/UCSC.jpg" title="bootRanges motivation" alt="bootRanges motivation" width="50%" style="display: block; margin: auto;"></p>
<p>Below are histograms from <span class="citation">Bickel et al.
(2010)</span> showing various ways of estimating the null distribution
of overlaps between two sequences, with the true null on top (known as
the sequences are generated from a random process, and so additional
sequences can be generated). The block bootstrap can do a much better
job of <span class="math inline">\(\color{brown}{\text{estimating the
variance of the null distribution}}\)</span>, compared to naive
permutation of features.</p>
<p><img src="images/comparsion.jpeg" title="Histogram comparison" alt="Histogram comparison" width="50%" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="overview-2">Overview<a class="anchor" aria-label="anchor" href="#overview-2"></a>
</h3>
<p>In a segmented block bootstrap, the blocks are sampled and placed
within regions of a genome <em>segmentation</em>. That is, for a genome
segmented into states 1,2,…,S, blocks from state <em>s</em> will be used
to tile the ranges of state <em>s</em> in each bootstrap sample. The
process can be visualized in (A), a block with length <span class="math inline">\(L_b\)</span> is <span class="math inline">\(\color{brown}{\text{randomly}}\)</span> selected
from state “red” and move to a <span class="math inline">\(\color{brown}{\text{tile}}\)</span> block across
chromosome within same states.</p>
<p>An example workflow of bootRanges used in combination with plyranges
is diagrammed in (B), and can be summarized as:</p>
<ol style="list-style-type: decimal">
<li>Overlap <em>GRanges</em> of feature <span class="math inline">\(x\)</span> and <em>GRanges</em> of feature <span class="math inline">\(y\)</span> to derive interested observed
statistics</li>
<li>
<code><a href="https://nullranges.github.io/nullranges/reference/bootRanges.html" class="external-link">bootRanges()</a></code> with optional <code>segmentation</code>
and <code>exclude</code> creates a <em>bootRanges</em> object <span class="math inline">\(y'\)</span>
</li>
<li>Overlap <em>GRanges</em> of feature <span class="math inline">\(x\)</span> and <span class="math inline">\(y'\)</span> to derive interested bootstrap
statistics</li>
<li>
<span class="math inline">\(z\)</span> test is performed for testing
the null hypothesis that there is no true biological enrichment</li>
</ol>
<p><img src="images/bootRanges.jpg" title="Overview of bootRanges" alt="Overview of bootRanges"></p>
</div>
<div class="section level3">
<h3 id="segmented-block-bootstrap">Segmented block bootstrap<a class="anchor" aria-label="anchor" href="#segmented-block-bootstrap"></a>
</h3>
<p>We demonstrate usage by loading a set of DNase hypersensitivity sites
(DHS) from the ENCODE project <span class="citation">(ENCODE
2012)</span> in A549 cell line (ENCSR614GWM). Here, for speed, we work
with a pre-processed data object from the nullrangesData package.</p>
<p>For speed of the vignette, we restrict to a smaller number of DHS,
filtering by the signal value. We also remove metadata columns that we
don’t need for the bootstrap analysis. Consider, when creating
bootstrapped data, that you will be creating an object many times larger
than your original features, so <span class="math inline">\(\color{brown}{\text{filtering and
trimming}}\)</span> extra metadata can help make the analysis more
efficient.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dhs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/nullrangesData/man/DHSA549Hg38.html" class="external-link">DHSA549Hg38</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?nullrangesData and browseVignettes('nullrangesData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="va">dhs</span> <span class="op">&lt;-</span> <span class="va">dhs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">signalValue</span> <span class="op">&gt;</span> <span class="fl">100</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="va">dhs</span></span>
<span><span class="co">#&gt; GRanges object with 6214 ranges and 1 metadata column:</span></span>
<span><span class="co">#&gt;          seqnames            ranges strand |        id</span></span>
<span><span class="co">#&gt;             &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;</span></span>
<span><span class="co">#&gt;      [1]     chr1     629126-629275      * |         1</span></span>
<span><span class="co">#&gt;      [2]     chr1     629306-629455      * |         2</span></span>
<span><span class="co">#&gt;      [3]     chr1     629901-630170      * |         3</span></span>
<span><span class="co">#&gt;      [4]     chr1     630346-630495      * |         4</span></span>
<span><span class="co">#&gt;      [5]     chr1     630521-630670      * |         5</span></span>
<span><span class="co">#&gt;      ...      ...               ...    ... .       ...</span></span>
<span><span class="co">#&gt;   [6210]     chrY 11307570-11307719      * |      6210</span></span>
<span><span class="co">#&gt;   [6211]     chrY 11308010-11308159      * |      6211</span></span>
<span><span class="co">#&gt;   [6212]     chrY 11320890-11321039      * |      6212</span></span>
<span><span class="co">#&gt;   [6213]     chrY 11322530-11322679      * |      6213</span></span>
<span><span class="co">#&gt;   [6214]     chrY 56834934-56835083      * |      6214</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 24 sequences from hg38 genome</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">dhs</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  chr1  chr2  chr3  chr4  chr5  chr6  chr7  chr8  chr9 chr10 chr11 chr12 chr13 </span></span>
<span><span class="co">#&gt;  1436   252   108    30   148    51   184   146   155   443   436   526    20 </span></span>
<span><span class="co">#&gt; chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22  chrX  chrY </span></span>
<span><span class="co">#&gt;   197   265   214   715    20   649   142    31    19    17    10</span></span></code></pre></div>
<div class="section level4">
<h4 id="import-excluded-regions">Import excluded regions<a class="anchor" aria-label="anchor" href="#import-excluded-regions"></a>
</h4>
<p><span class="math inline">\(\color{brown}{\text{To avoid placing
bootstrap features into regions of the genome that don’t typically have
features}}\)</span>. We next import excluded regions including
ENCODE-produced excludable regions<span class="citation">(Amemiya,
Kundaje, and Boyle 2019)</span>, telomeres from UCSC, centromeres <span class="citation">(Commo 2022)</span>. For easy use, pre-combined
excludable regions is stored in <em>ExperimentHub.</em></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">eh</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># query(eh, "nullrangesdata")</span></span>
<span><span class="va">exclude</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH7306"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="pre-built-segmentations">Pre-built segmentations<a class="anchor" aria-label="anchor" href="#pre-built-segmentations"></a>
</h4>
<p>The segmented block bootstrap has two options, either:</p>
<ul>
<li>Perform a de-novo segmentation of the genome using feature density,
e.g. gene density</li>
<li>Use exiting segmentation (e.g. ChromHMM, etc.) downloaded from
AnnotationHub or external to Bioconductor (BED files imported with
<em>rtracklayer</em>)</li>
</ul>
<p><code><a href="https://nullranges.github.io/nullranges/reference/segmentDensity.html" class="external-link">segmentDensity()</a></code> from <em>nullranges</em> provide first
option by <em>CBS</em> or <em>HMM</em> methods. Here we loaded pre-built
segmentations for easy use in hg38 with <span class="math inline">\(L_s=2e6\)</span> considering excludable regions.
Assessment of segmentation through ranges plot, bar plot and box plot
could be plotted by <code><a href="https://nullranges.github.io/nullranges/reference/plotSegment.html" class="external-link">plotSegment()</a></code>. Detailed Usage is shown
in vignette <a href="https://nullranges.github.io/nullranges/articles/segmented_boot_ranges.html" class="external-link">Segmented
block bootstrap</a> from the nullranges package.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seg_cbs</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH7307"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="subsampling-with-bootranges">Subsampling with bootRanges()<a class="anchor" aria-label="anchor" href="#subsampling-with-bootranges"></a>
</h4>
<p>Now we apply a segmented block bootstrap with blocks of size 500kb,
to the peaks. Here we show generation of 10 iterations of a block
bootstrap followed by a typical overlap analysis using
<em>plyranges</em> <span class="citation">(Lee, Cook, and Lawrence
2019)</span>. (We might normally do 100 iterations or more, depending on
the granularity of the bootstrap distribution that is needed.)</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span> <span class="co"># for reproducibility</span></span>
<span><span class="va">R</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span><span class="va">blockLength</span> <span class="op">&lt;-</span> <span class="fl">5e5</span></span>
<span><span class="va">boots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://nullranges.github.io/nullranges/reference/bootRanges.html" class="external-link">bootRanges</a></span><span class="op">(</span><span class="va">dhs</span>, <span class="va">blockLength</span>, R <span class="op">=</span> <span class="va">R</span>, seg <span class="op">=</span> <span class="va">seg_cbs</span>, exclude<span class="op">=</span><span class="va">exclude</span><span class="op">)</span></span>
<span><span class="va">boots</span></span>
<span><span class="co">#&gt; bootRanges object with 62764 ranges and 4 metadata columns:</span></span>
<span><span class="co">#&gt;           seqnames            ranges strand |        id     block  iter</span></span>
<span><span class="co">#&gt;              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt; &lt;Rle&gt;</span></span>
<span><span class="co">#&gt;       [1]     chr1     242791-242940      * |       347         5     1</span></span>
<span><span class="co">#&gt;       [2]     chr1     256031-256180      * |       348         5     1</span></span>
<span><span class="co">#&gt;       [3]     chr1     391535-391684      * |      5301         8     1</span></span>
<span><span class="co">#&gt;       [4]     chr1     421046-421195      * |      5302         8     1</span></span>
<span><span class="co">#&gt;       [5]     chr1     438186-438335      * |      5303         8     1</span></span>
<span><span class="co">#&gt;       ...      ...               ...    ... .       ...       ...   ...</span></span>
<span><span class="co">#&gt;   [62760]     chrY 22010887-22011036      * |      6075     12428    10</span></span>
<span><span class="co">#&gt;   [62761]     chrY 25408146-25408295      * |      1826     12436    10</span></span>
<span><span class="co">#&gt;   [62762]     chrY 26573840-26573989      * |       576     12439    10</span></span>
<span><span class="co">#&gt;   [62763]     chrY 26595240-26595389      * |       577     12439    10</span></span>
<span><span class="co">#&gt;   [62764]     chrY 28050177-28050326      * |      3027     12443    10</span></span>
<span><span class="co">#&gt;           blockLength</span></span>
<span><span class="co">#&gt;                 &lt;Rle&gt;</span></span>
<span><span class="co">#&gt;       [1]      500000</span></span>
<span><span class="co">#&gt;       [2]      500000</span></span>
<span><span class="co">#&gt;       [3]      500000</span></span>
<span><span class="co">#&gt;       [4]      500000</span></span>
<span><span class="co">#&gt;       [5]      500000</span></span>
<span><span class="co">#&gt;       ...         ...</span></span>
<span><span class="co">#&gt;   [62760]      500000</span></span>
<span><span class="co">#&gt;   [62761]      500000</span></span>
<span><span class="co">#&gt;   [62762]      500000</span></span>
<span><span class="co">#&gt;   [62763]      500000</span></span>
<span><span class="co">#&gt;   [62764]      500000</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 24 sequences from hg38 genome</span></span></code></pre></div>
<p>What is returned here? The <code>bootRanges</code> function returns a
<em>bootRanges</em> object, which is a simple sub-class of
<em>GRanges</em>. The iteration (<code>iter</code>) and the block length
(<code>blockLength</code>) are recorded as metadata columns, accessible
via <code>mcols</code>.</p>
</div>
</div>
<div class="section level3">
<h3 id="assessing-quality-of-subsampling">Assessing quality of subsampling<a class="anchor" aria-label="anchor" href="#assessing-quality-of-subsampling"></a>
</h3>
<p>We can examine properties of permuted y over iterations, and compare
to the original y. To do so, we first add the original features as
iter=0. Then compute summaries:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="va">dhs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>iter<span class="op">=</span><span class="fl">0</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-bind.html" class="external-link">bind_ranges</a></span><span class="op">(</span><span class="va">boots</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">iter</span><span class="op">)</span></span>
<span><span class="va">stats</span> <span class="op">&lt;-</span> <span class="va">combined</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">stats</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">#&gt;   iter      n</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 0      <span style="text-decoration: underline;">6</span>214</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 1      <span style="text-decoration: underline;">6</span>149</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 2      <span style="text-decoration: underline;">6</span>275</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 3      <span style="text-decoration: underline;">6</span>249</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 4      <span style="text-decoration: underline;">6</span>293</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 5      <span style="text-decoration: underline;">6</span>365</span></span></code></pre></div>
<p>We can also look at distributions of various aspects, e.g. here the
inter-feature distance of features, across a few of the bootstraps and
the original feature set y.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/ggridges/" class="external-link">ggridges</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org" class="external-link">purrr</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">interdist</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">dat</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">x</span> <span class="op">=</span> <span class="va">dat</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span>,<span class="op">]</span></span>
<span>    <span class="va">y</span> <span class="op">=</span> <span class="va">dat</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span>,<span class="op">]</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">seqnames</span> <span class="op">==</span> <span class="va">y</span><span class="op">$</span><span class="va">seqnames</span>, <span class="va">x</span><span class="op">$</span><span class="va">start</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">width</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span> <span class="op">-</span> <span class="va">y</span><span class="op">$</span><span class="va">start</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">width</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">}</span></span>
<span></span>
<span><span class="va">combined</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">iter</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fl">0</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">nest</a></span><span class="op">(</span><span class="op">-</span><span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>interdist <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">data</span>, <span class="op">~</span><span class="fu">interdist</span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">iter</span>,<span class="va">interdist</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">unnest</a></span><span class="op">(</span><span class="va">interdist</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">iter</span> <span class="op">==</span> <span class="fl">0</span>, <span class="st">"original"</span>, <span class="st">"boot"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">interdist</span><span class="op">)</span>, <span class="va">iter</span>, fill<span class="op">=</span><span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://wilkelab.org/ggridges/reference/geom_density_ridges.html" class="external-link">geom_density_ridges</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>data<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">stats</span>,<span class="fl">4</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="fl">1.5</span>, y<span class="op">=</span><span class="va">iter</span>, label<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"n="</span>,<span class="va">n</span><span class="op">)</span>, fill<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span>,</span>
<span>            vjust<span class="op">=</span><span class="fl">1.5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.198</span></span></code></pre></div>
<p><img src="nullranges_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="derive-interested-statistics">Derive interested statistics<a class="anchor" aria-label="anchor" href="#derive-interested-statistics"></a>
</h3>
<p>Suppose we have a set of features <code>x</code> and we are
interested in</p>
<p>evaluating the enrichment of this set with the DHS. We can calculate
for example the sum observed number of overlaps for features in
<code>x</code> with DHS in whole genome (or something more complicated,
e.g. the maximum log fold change or signal value for DHS peaks within a
<code>maxgap</code> window of <code>x</code>).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr2"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="fl">50</span><span class="op">:</span><span class="fl">99</span> <span class="op">*</span> <span class="fl">1e6</span>, width<span class="op">=</span><span class="fl">1e6</span><span class="op">)</span>, x_id<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">x</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>n_overlaps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-count-overlaps.html" class="external-link">count_overlaps</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">dhs</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span> <span class="va">x</span><span class="op">$</span><span class="va">n_overlaps</span> <span class="op">)</span></span>
<span><span class="co">#&gt; [1] 64</span></span></code></pre></div>
<p>We can repeat this with the bootstrapped features using a
<code>group_by</code> command, a <code>summarize</code>, followed by a
second <code>group_by</code> and <code>summarize</code>. It may help to
step through these commands one by one to understand what the
intermediate output is.</p>
<p>Note that we need to use <code><a href="https://tidyr.tidyverse.org/reference/complete.html" class="external-link">tidyr::complete</a></code> in order to
fill in combinations of <code>x</code> and <code>iter</code> where the
overlap was 0.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">boot_stats</span> <span class="op">&lt;-</span> <span class="va">x</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_inner</a></span><span class="op">(</span><span class="va">boots</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>n_overlaps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/complete.html" class="external-link">complete</a></span><span class="op">(</span><span class="va">iter</span>, fill<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>n_overlaps <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>meanOverlaps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">n_overlaps</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>If one is interested in assessing a <span class="math inline">\(\color{brown}{\text{feature-wise}}\)</span>
statistics instead of <span class="math inline">\(\color{brown}{\text{genome-wise}}\)</span>
statistics, eg.,the mean observed number of overlaps per feature or mean
base pair overlap in <code>x</code>, one can also group by both
(<code>block</code>,<code>iter</code>). 10,000 total blocks may
therefore be sufficient to derive a bootstrap distribution, avoiding the
need to generate many bootstrap genomes of data.</p>
<p>Finally we can plot a histogram. In this case, as the <code>x</code>
features were arbitrary, our observed value falls within the
distribution of mean overlap with bootstrapped data.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">boot_stats</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">meanOverlaps</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code></pre></div>
<p><img src="nullranges_files/figure-html/boot-hist-1.png" width="700"></p>
<p>For more examples of combining <code>bootRanges</code> from
<em>nullranges</em> with <em>plyranges</em> piped operations, see the
relevant chapter in the <a href="https://nullranges.github.io/tidy-ranges-tutorial/bootstrap-overlap.html" class="external-link">tidy-ranges-tutorial</a>
book.</p>
</div>
<div class="section level3">
<h3 id="extension">Extension<a class="anchor" aria-label="anchor" href="#extension"></a>
</h3>
<div class="section level4">
<h4 id="optimized-effect-size-threshold">Optimized effect size threshold<a class="anchor" aria-label="anchor" href="#optimized-effect-size-threshold"></a>
</h4>
<p>Penalized splines across a range of effect sizes on the null sets can
be modeled, through which to derive confidence interval at the same time
on every effect size. Therefore, an optimized effect size threshold
could be derived, eg. rather than limiting DEG logFC at a arbitrary
threshold.</p>
<p><img src="images/effectSize.jpeg" title="effect size" alt="effect size" width="50%" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="derive-statistics-of-interest-from-other-formats">Derive statistics of interest from other formats<a class="anchor" aria-label="anchor" href="#derive-statistics-of-interest-from-other-formats"></a>
</h4>
<p>Instead of deriving statistics of interest from <em>GRanges</em>
metadata column, count matrix from <em>SummerizedExperiment</em> or
<em>SingleCellExperiment</em> could also be used. One case study is to
assess the correlation of the all pairs of genes and promoter peaks from
Chromium Single Cell Multiome ATAC + Gene Expression.</p>
<ul>
<li>Plyrangs pseudo code saving the count matrix in GRanges’s metadata
column as a <code><a href="https://rdrr.io/pkg/IRanges/man/AtomicList-class.html" class="external-link">NumericList()</a></code> format and use <em>Plyranges</em>
in downstream analysis.</li>
</ul>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## split sparse count matrix into NumericList</span></span>
<span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="va">rna_Granges</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">rna.sd</span><span class="op">==</span><span class="fl">0</span><span class="op">)</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>counts1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/AtomicList-class.html" class="external-link">NumericList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/asplit.html" class="external-link">asplit</a></span><span class="op">(</span><span class="va">rna.scaled</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">promoter</span> <span class="op">&lt;-</span> <span class="va">promoter_Granges</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">promoter.sd</span><span class="op">==</span><span class="fl">0</span><span class="op">)</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>counts2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/AtomicList-class.html" class="external-link">NumericList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/asplit.html" class="external-link">asplit</a></span><span class="op">(</span><span class="va">promoter.scaled</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bootranges</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://nullranges.github.io/nullranges/reference/bootRanges.html" class="external-link">bootRanges</a></span><span class="op">(</span><span class="va">promoter</span>,blockLength <span class="op">=</span> <span class="fl">5e5</span>,R<span class="op">=</span><span class="va">R</span>,type <span class="op">=</span> <span class="st">"bootstrap"</span>, withinChrom <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co">## draw mean correlation distribution plot</span></span>
<span><span class="va">cor_whole</span><span class="op">&lt;-</span><span class="va">rna</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_inner</a></span><span class="op">(</span><span class="va">bootranges</span>, maxgap<span class="op">=</span><span class="fl">1000</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>rho <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html" class="external-link">n</a></span><span class="op">(</span><span class="va">col</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">counts1</span> <span class="op">*</span> <span class="va">counts2</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">rho</span>,<span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>meanCor <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">rho</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>TidySE pesudo code</li>
</ul>
<p>Or creating a <em>SummarizedExperiment</em> object and use
<em>tidySummarizedExperiment</em> to derive statistics of interest.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tidySummarizedExperiment</span><span class="op">)</span></span>
<span><span class="co">## make an SE where each row is an overlap</span></span>
<span><span class="va">se_rna</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="op">(</span></span>
<span>  assays<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>rna<span class="op">=</span><span class="va">rna.scaled</span><span class="op">)</span>,</span>
<span>  rowRanges<span class="op">=</span><span class="va">rna_Granges</span><span class="op">)</span></span>
<span><span class="va">se_promoter</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="op">(</span></span>
<span>  assays<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>promoter<span class="op">=</span><span class="va">promoter.scaled</span><span class="op">)</span>,</span>
<span>  rowRanges<span class="op">=</span><span class="va">promoter_Granges</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## make an SE where each row is an overlap</span></span>
<span><span class="va">makeOverlapSE</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">se_rna</span>, <span class="va">se_promoter</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu">rowRanges</span><span class="op">(</span><span class="va">se_rna</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_inner</a></span><span class="op">(</span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">se_promoter</span><span class="op">)</span>,maxgap <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span>  <span class="va">assay_x</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">se_rna</span>, <span class="st">"rna"</span><span class="op">)</span><span class="op">[</span> <span class="va">idx</span><span class="op">$</span><span class="va">gene</span>, <span class="op">]</span></span>
<span>  <span class="va">assay_y</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">se_promoter</span>, <span class="st">"promoter"</span><span class="op">)</span><span class="op">[</span> <span class="va">idx</span><span class="op">$</span><span class="va">peak</span>, <span class="op">]</span></span>
<span>  <span class="co"># this is needed to build the SE</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">assay_x</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">assay_y</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span> <span class="va">idx</span><span class="op">$</span><span class="va">gene</span> <span class="op">)</span></span>
<span>  <span class="va">new_ranges</span> <span class="op">&lt;-</span> <span class="fu">rowRanges</span><span class="op">(</span><span class="va">se_rna</span><span class="op">)</span><span class="op">[</span> <span class="va">idx</span><span class="op">$</span><span class="va">gene</span> <span class="op">]</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">new_ranges</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span> <span class="va">idx</span><span class="op">$</span><span class="va">gene</span> <span class="op">)</span></span>
<span>  <span class="fu">SummarizedExperiment</span><span class="op">(</span></span>
<span>    assays<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">assay_x</span>, y<span class="op">=</span><span class="va">assay_y</span><span class="op">)</span>,</span>
<span>    rowRanges<span class="op">=</span><span class="va">new_ranges</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">makeOverlapSE</span><span class="op">(</span><span class="va">se_rna</span>, <span class="va">se_promoter</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>rho <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">data</span>,</span>
<span>                   <span class="kw">function</span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>rho <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">unnest</a></span><span class="op">(</span><span class="va">rho</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="additional-references">Additional references<a class="anchor" aria-label="anchor" href="#additional-references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-encode_exclude" class="csl-entry">
Amemiya, Haley M, Anshul Kundaje, and Alan P Boyle. 2019. <span>“The
<span>ENCODE</span> Blacklist: Identification of Problematic Regions of
the Genome.”</span> <em>Scientific Reports</em> 9 (1): 9354. <a href="https://doi.org/10.1038/s41598-019-45839-z" class="external-link">https://doi.org/10.1038/s41598-019-45839-z</a>.
</div>
<div id="ref-bickel_2010" class="csl-entry">
Bickel, Peter J., Nathan Boley, James B. Brown, Haiyan Huang, and Nancy
R. Zhang. 2010. <span>“Subsampling Methods for Genomic
Inference.”</span> <em>The Annals of Applied Statistics</em> 4 (4):
1660–97. <a href="https://doi.org/10.1214/10-%7BAOAS363%7D" class="external-link">https://doi.org/10.1214/10-{AOAS363}</a>.
</div>
<div id="ref-rCGH" class="csl-entry">
Commo, Frederic. 2022. <em>rCGH: Comprehensive Pipeline for Analyzing
and Visualizing Array-Based CGH Data</em>. <a href="https://github.com/fredcommo/rCGH" class="external-link">https://github.com/fredcommo/rCGH</a>.
</div>
<div id="ref-de_2014" class="csl-entry">
De, Subhajyoti, Brent S Pedersen, and Katerina Kechris. 2014. <span>“The
Dilemma of Choosing the Ideal Permutation Strategy While Estimating
Statistical Significance of Genome-Wide Enrichment.”</span>
<em>Briefings in Bioinformatics</em> 15 (6): 919–28. <a href="https://doi.org/10.1093/bib/bbt053" class="external-link">https://doi.org/10.1093/bib/bbt053</a>.
</div>
<div id="ref-encode" class="csl-entry">
ENCODE. 2012. <span>“<span class="nocase">An integrated encyclopedia of
DNA elements in the human genome</span>.”</span> <em>Nature</em> 489
(7414): 57–74. <a href="https://doi.org/10.1038/nature11247" class="external-link">https://doi.org/10.1038/nature11247</a>.
</div>
<div id="ref-GenometriCorr" class="csl-entry">
Favorov, Alexander, Loris Mularoni, Leslie M Cope, Yulia Medvedeva,
Andrey A Mironov, Vsevolod J Makeev, and Sarah J Wheelan. 2012.
<span>“Exploring Massive, Genome Scale Datasets with the
<span>GenometriCorr</span> Package.”</span> <em><span>PLoS</span>
Computational Biology</em> 8 (5): e1002529. <a href="https://doi.org/10.1371/journal.pcbi.1002529" class="external-link">https://doi.org/10.1371/journal.pcbi.1002529</a>.
</div>
<div id="ref-regioneR" class="csl-entry">
Gel, Bernat, Anna Díez-Villanueva, Eduard Serra, Marcus Buschbeck,
Miguel A Peinado, and Roberto Malinverni. 2016. <span>“<span class="nocase">regioneR</span>: An r/Bioconductor Package for the
Association Analysis of Genomic Regions Based on Permutation
Tests.”</span> <em>Bioinformatics</em> 32 (2): 289–91. <a href="https://doi.org/10.1093/bioinformatics/btv562" class="external-link">https://doi.org/10.1093/bioinformatics/btv562</a>.
</div>
<div id="ref-haiminen_2007" class="csl-entry">
Haiminen, Niina, Heikki Mannila, and Evimaria Terzi. 2007.
<span>“Comparing Segmentations by Applying Randomization
Techniques.”</span> <em><span>BMC</span> Bioinformatics</em> 8 (May):
171. <a href="https://doi.org/10.1186/1471-2105-8-171" class="external-link">https://doi.org/10.1186/1471-2105-8-171</a>.
</div>
<div id="ref-GAT" class="csl-entry">
Heger, Andreas, Caleb Webber, Martin Goodson, Chris P Ponting, and
Gerton Lunter. 2013. <span>“<span>GAT</span>: A Simulation Framework for
Testing the Association of Genomic Intervals.”</span>
<em>Bioinformatics</em> 29 (16): 2046–48. <a href="https://doi.org/10.1093/bioinformatics/btt343" class="external-link">https://doi.org/10.1093/bioinformatics/btt343</a>.
</div>
<div id="ref-huen_2010" class="csl-entry">
Huen, David S, and Steven Russell. 2010. <span>“On the Use of Resampling
Tests for Evaluating Statistical Significance of Binding-Site
Co-Occurrence.”</span> <em><span>BMC</span> Bioinformatics</em> 11
(June): 359. <a href="https://doi.org/10.1186/1471-2105-11-359" class="external-link">https://doi.org/10.1186/1471-2105-11-359</a>.
</div>
<div id="ref-kanduri_2019" class="csl-entry">
Kanduri, Chakravarthi, Christoph Bock, Sveinung Gundersen, Eivind Hovig,
and Geir Kjetil Sandve. 2019. <span>“Colocalization Analyses of Genomic
Elements: Approaches, Recommendations and Challenges.”</span>
<em>Bioinformatics</em> 35 (9): 1615–24. <a href="https://doi.org/10.1093/bioinformatics/bty835" class="external-link">https://doi.org/10.1093/bioinformatics/bty835</a>.
</div>
<div id="ref-granges" class="csl-entry">
Lawrence, Wolfgang AND Pagès, Michael AND Huber. 2013. <span>“Software
for Computing and Annotating Genomic Ranges.”</span> <em>PLOS
Computational Biology</em> 9 (8): 1–10. <a href="https://doi.org/10.1371/journal.pcbi.1003118" class="external-link">https://doi.org/10.1371/journal.pcbi.1003118</a>.
</div>
<div id="ref-Lee2019" class="csl-entry">
Lee, Stuart, Dianne Cook, and Michael Lawrence. 2019. <span>“Plyranges:
A Grammar of Genomic Data Transformation.”</span> <em>Genome
Biology</em> 20 (1): 4. <a href="https://doi.org/10.1186/s13059-018-1597-8" class="external-link">https://doi.org/10.1186/s13059-018-1597-8</a>.
</div>
<div id="ref-GREAT" class="csl-entry">
McLean, Cory Y., Dave Bristor, Michael Hiller, Shoa L. Clarke, Bruce T.
Schaar, Craig B. Lowe, Aaron M. Wenger, and Gill Bejerano. 2010.
<span>“GREAT Improves Functional Interpretation of Cis-Regulatory
Regions.”</span> <em>Nature Biotechnology</em> 28 (5): 495–501. <a href="https://doi.org/10.1038/nbt.1630" class="external-link">https://doi.org/10.1038/nbt.1630</a>.
</div>
<div id="ref-LOLA" class="csl-entry">
Sheffield, Nathan C, and Christoph Bock. 2016. <span>“<span>LOLA</span>:
Enrichment Analysis for Genomic Region Sets and Regulatory Elements in r
and Bioconductor.”</span> <em>Bioinformatics</em> 32 (4): 587–89. <a href="https://doi.org/10.1093/bioinformatics/btv612" class="external-link">https://doi.org/10.1093/bioinformatics/btv612</a>.
</div>
<div id="ref-ChIP-Enrich" class="csl-entry">
Welch, Ryan P., Chee Lee, Paul M. Imbriano, Snehal Patil, Terry E.
Weymouth, R. Alex Smith, Laura J. Scott, and Maureen A. Sartor. 2014.
<span>“<span class="nocase">ChIP-Enrich: gene set enrichment testing for
ChIP-seq data</span>.”</span> <em>Nucleic Acids Research</em> 42 (13):
e105–5. <a href="https://doi.org/10.1093/nar/gku463" class="external-link">https://doi.org/10.1093/nar/gku463</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wancen Mu, Eric Davis, Mikhail Dozmorov, Stuart Lee, Michael Love, Douglas Phanstiel.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
